File: corpus/corpus.ts
import * as fs from 'fs';

// Define valid characters as a string
const validChars = "abcdefghijklmnopqrstuvwxyz ".split('');

// Function to find intersection of two lists
function intersection(lst1: string[], lst2: string[]): string[] {
    const temp = new Set(lst2);
    return lst1.filter(value => temp.has(value));
}

// Function to find n-grams (bigrams, etc.) in a string
function findNgrams(str: string, n: number): [string, string][] {
    const inputList = intersection(str.toLowerCase().split(''), validChars);
    const result: [string, string][] = [];
    for (let i = 0; i < inputList.length - n + 1; i++) {
        result.push(inputList.slice(i, i + n) as [string, string]);
    }
    return result;
}

// Function to calculate unigrams
function unigram(language: string, file: string): Record<string, number> {
    const text = fs.readFileSync(file, 'utf-8').toLowerCase();
    const occurrence = new Map<string, number>();
    const filteredText = intersection(text.split(''), validChars);

    filteredText.forEach((char) => {
        occurrence.set(char, (occurrence.get(char) || 0) + 1);
    });

    const count = filteredText.length;
    const probabilities: Record<string, number> = {};

    validChars.forEach((char) => {
        const prob = (occurrence.get(char) || 0) / count;
        probabilities[char] = parseFloat(prob.toExponential(6));
    });

    console.log("unigram", language, file, count);
    return probabilities;
}


function bigram(language: string, file: string): Record<string, number> {
    const text = fs.readFileSync(file, 'utf-8').toLowerCase();
    const occurrence = new Map<string, number>();
    const bigrams = findNgrams(text, 2);

    bigrams.forEach(([char1, char2]) => {
        const key = char1 + char2;
        occurrence.set(key, (occurrence.get(key) || 0) + 1);
    });

    const count = bigrams.length;
    const probabilities: Record<string, number> = {};

    validChars.forEach(char1 => {
        validChars.forEach(char2 => {
            const key = char1 + char2;
            const prob = (occurrence.get(key) || 0) / count;
            probabilities[key] = parseFloat(prob.toExponential(6));
        });
    });

    console.log("bigram", language, file, count);
    return probabilities; // Return the JSON object
}

// Corpus of languages and corresponding files
const corpus: { [key: string]: string } = {
    "danish": "corpus/dan_news_2008_30K-sentences.txt",
    "english": "corpus/eng_news_2016_30K-sentences.txt",
    "french": "corpus/fra_news_2010_30K-sentences.txt",
    "german": "corpus/deu_news_2015_30K-sentences.txt",
    "swedish": "corpus/swe_news_2007_30K-sentences.txt",
};

function saveJsonToFile(filename: string, jsonData: Record<string, number>) {
    fs.writeFileSync(filename, JSON.stringify(jsonData, null, 2), 'utf-8');
}

// Loop over each language and process bigrams (unigram can be called similarly)
for (const language in corpus) {
    saveJsonToFile(`website/assets/unigram_${language}.json`, unigram(language, corpus[language]));
    saveJsonToFile(`website/assets/bigram_${language}.json`, bigram(language, corpus[language]));
}

File: corpus/corpus.py
import json
from collections import Counter

# Define valid characters as a set for faster lookup
VALID_CHARS = set("abcdefghijklmnopqrstuvwxyz ")

def filter_valid_chars(text):
    """Filters a string to include only valid characters."""
    return [char for char in text.lower() if char in VALID_CHARS]

def find_ngrams(text, n):
    """Finds n-grams (bigrams, trigrams, etc.) in a string."""
    filtered_text = filter_valid_chars(text)
    return [tuple(filtered_text[i:i + n]) for i in range(len(filtered_text) - n + 1)]

def unigram(language, filepath):
    """Computes unigram probabilities from a text file."""
    with open(filepath, 'r', encoding='utf-8') as f:
        text = f.read()
    
    filtered_text = filter_valid_chars(text)
    total_count = len(filtered_text)
    occurrence = Counter(filtered_text)
    
    probabilities = {
        char: round(occurrence.get(char, 0) / total_count, 6)
        for char in VALID_CHARS
    }

    print(f"Unigram processed: {language} ({total_count} characters)")
    return probabilities

def bigram(language, filepath):
    """Computes bigram probabilities from a text file."""
    with open(filepath, 'r', encoding='utf-8') as f:
        text = f.read()

    bigrams = find_ngrams(text, 2)
    total_count = len(bigrams)
    occurrence = Counter([''.join(bigram) for bigram in bigrams])

    probabilities = {
        f"{char1}{char2}": round(occurrence.get(f"{char1}{char2}", 0) / total_count, 6)
        for char1 in VALID_CHARS
        for char2 in VALID_CHARS
    }

    print(f"Bigram processed: {language} ({total_count} bigrams)")
    return {
        "total_count": total_count,
        "probabilities": probabilities
    }

def save_json(filename, data):
    """Saves dictionary data to a JSON file."""
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2)

# Corpus files
CORPUS = {
    "en": "corpus/en_sentences.txt",
    "es": "corpus/es_sentences.txt",
    "de": "corpus/de_sentences.txt",
    "fr": "corpus/fr_sentences.txt",
    "it": "corpus/it_sentences.txt",
    "nl": "corpus/nl_sentences.txt",
    "pt": "corpus/pt_sentences.txt",
    "da": "corpus/da_sentences.txt",
    "sv": "corpus/sv_sentences.txt",
    "no": "corpus/no_sentences.txt",
    "fi": "corpus/fi_sentences.txt",
    "ro": "corpus/ro_sentences.txt",
    "hu": "corpus/hu_sentences.txt",
    "tr": "corpus/tr_sentences.txt",
    "id": "corpus/id_sentences.txt",
}

if __name__ == "__main__":
    for lang, file in CORPUS.items():
        save_json(f"resources/data/unigram_{lang}.json", unigram(lang, file))
        save_json(f"resources/data/bigram_{lang}.json", bigram(lang, file))
